README.md

```diff
@@ -7,10 +7,30 @@ The best way to use it is via [vaadin.com/start](https://vaadin.com/start) - you
 
 To access it directly from github, clone the repository and import the project to the IDE of your choice as a Maven project. You need to have Java 8 or 11 installed.
 
-Run using `mvn -Pprepare-osgi-container -Prun-osgi-container -Pproduction verify` and open [http://localhost:8080](http://localhost:8080) in browser.
-This command will run profiles which downloads for you Felix OSGi container and installs `felix-jetty` 
-as OSGi web container. It deploys all necessary bundles to Felix and start it synchronously.
-The container will be stopped  when you stop your maven build (e.g. by killing it via `Ctrl+C`).
+Main Command: `mvn clean install`
+- build all bundles
+- run tests inside the osgi framework when `tests.bndrun` file exists in project.
+- generates the Application if `application.bndrun` file exists in project.
+
+Run the Application using `java -jar application/application.jar`
```

Comment:
Since `install` command fails `application.jar` is not generated.

Reply:
works now! I fixed the path in the docs.  added (`target`) folder
`java -jar application/target/application.jar`

Reply:
The main reason why it failed for me is: `httpclient-4.5.13.jar` is included in the final application jar instead of the correct one `httpclient-osgi.jar`.

You have excluded explicitly  the artifact 
```
<groupId>org.apache.httpcomponents</groupId>
<artifactId>httpclient</artifactId>
```
from the build.
But :
- this is not required since this artifact is not an OSGi bundle ( so it won't affect anything )
- it doesn't help to me anyhow: the artifact is still included
- what's more important the valid OSGi bundle `httpclient-osgi.jar` is never included into the produced application jar.

As a result : the app doesn't work just because main `flow-server`  bundle dependencies are not installed , as a result `flow-server`  is not active and nothing works in OSGi at all.

I have not checked your latest changes but I'm pretty sure that it has not improved anyhow.
In my project which I create from scratch following your way the issue is exactly the same.

It seems that `resolve`  goal is not able to resolve the artifact properly because both jars `httpclient-osgi.jar` and `httpclient.jar` contains the same packages and `bnd-resolver:resolve` simply fails to select the correct jar.

I'm trying to resolve this issue now.

Reply:
https://github.com/bndtools/bnd/issues/4380

Reply:
I've made httpclient bundle optional in Flow.
Now I'm able to build and run the app without unsatisfied dependencies.

But HTTP doesn't work : https://github.com/vaadin/base-starter-flow-osgi/pull/97

Reply:
HTTP Whiteboard is working
http://localhost:8080/icons/icon.png


But VaadinServlet.service has a problem
```
javax.servlet.ServletException: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:254)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
	at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
	at org.apache.felix.http.base.internal.dispatch.Dispatcher$1.doFilter(Dispatcher.java:146)
	at org.apache.felix.http.base.internal.whiteboard.WhiteboardManager.invokePreprocessors(WhiteboardManager.java:988)
	at org.apache.felix.http.base.internal.dispatch.Dispatcher.dispatch(Dispatcher.java:91)
	at org.apache.felix.http.base.internal.dispatch.DispatcherServlet.service(DispatcherServlet.java:49)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:569)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1377)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:507)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1292)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
	at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1617)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1580)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:252)
	... 33 more
Caused by: java.lang.NullPointerException
	at com.vaadin.flow.server.frontend.FrontendUtils.getResourceFromWebpack(FrontendUtils.java:420)
	at com.vaadin.flow.server.frontend.FrontendUtils.getStatsAssetsByChunkName(FrontendUtils.java:528)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.appendNpmBundle(BootstrapHandler.java:904)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupFrameworkLibraries(BootstrapHandler.java:886)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupDocumentHead(BootstrapHandler.java:765)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.getBootstrapPage(BootstrapHandler.java:539)
	at com.vaadin.flow.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHan
```

Reply:
>HTTP Whiteboard is working
>http://localhost:8080/icons/icon.png

True.
The icon is shown.


>But VaadinServlet.service has a problem

My bad.

To be able to run project correctly one has to :
- go to the `starter`  module and run : `mvn install -Pproduction`. `production` profile is important because it runs webpack which creates the stats file and then there will be no the exception above.
- then go to the `app`  and run `mvn install`.
- then run the app from this module as `java -jar target/app.jar`.

I don't see any exception when I access `http://localhost:8080` but nothing is shown. The page gives me 404.

Reply:
ah production mone.. ok!

Use this:
```
mvn install -Pproduction
mvn bnd-resolver:resolve -f app/pom.xml 
mvn bnd-run:run -f app/pom.xml 

```

Reply:
Now i tried it also with your VaadinServletRegistration.

this is only working when

- adding a reference to ResourceProvider

```
@Reference(cardinality = ReferenceCardinality.MANDATORY)
ResourceProvider pr;
```

- chnng bundle order after resolving

Reply:
Hmmm....
Interesting.
So it works for you...

It seems there should be some exceptions.
But I don't see any.
And 404 should not be returned on this case. But rather 503.
I really dislike how things work here: no any way to understand what's wrong.
But I will check tomorrow how it behaves with a reference to the `ResourceProvider`.

Anyway: I have to say thank you very much for you help even though I still have problems your help is very valuable !

Reply:
ah sorry i had this .
i had this in my base. 

```
-runrequires.flow: \
	bnd.identity;version='2.5.0';id='com.vaadin.flow.client',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.data',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.html.components',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.push'
```

this is what runs in my computer.
https://github.com/bipolis/base-starter-flow-osgi/tree/feature/osgi-bnd

Changes 
https://github.com/bipolis/base-starter-flow-osgi/commit/801961abcc10bbc8bf14cdaa16b5e4e9f445fd61


Not all nessecary bundles are chatched from resolver.

Reply:
Prevent activation of a VaadinServlet before ResourceProvider exists.
Q: What starts the activation of a VaadinServlet?
A: VaadinServletRegistration is `immediate=true`. - will activate on bundle-start
So i removed VaadinServletRegistration and used only the FixedVaadinServlet. `immediate=false` by default.

Q: What activate the FixedVaadinServlet
A: HttpWhiteboard looks for every Servlet and activates, if the Servlet matches the Servletcontext. This is when the Bundle of the FixedVaadinServlet is started AND the Whiteboards will be initialized. does not help enough- could be before ResourceProvider is started.

So i bound the FixedVaadinServlet to an other new ServletContext then the Default one.
https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.http.whiteboard.html#service.http.whiteboard.common.properties
```
@Component(service = { Servlet.class
		// Same like @VaadinContext
		// , property = {HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_SELECT + "=(" +
		// HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_NAME +"=vaadin")"})

})
@VaadinMode
@VaadinContext
@HttpWhiteboardServletAsyncSupported
@HttpWhiteboardServletPattern("/*")
public class FixedVaadinServlet extends VaadinServlet {

```
Now this Servlet would not be activates from HttpWhiteboard with the DefaultContext

Log:
```
[DEBUG] [ServiceReference 30 from bundle 2 : com.example.osgi.base:1.0.0 ref=[javax.servlet.Servlet] properties={component.id=0, component.name=com.example.starter.base.osgi.FixedVaadinServlet, objectClass=[javax.servlet.Servlet], osgi.http.whiteboard.context.select=(osgi.http.whiteboard.context.name=vaadin), osgi.http.whiteboard.servlet.asyncSupported=true, osgi.http.whiteboard.servlet.pattern=[/*], service.bundleid=2, service.id=30, service.scope=bundle, servlet.init.compatibilityMode=false}] Ignoring unmatching Servlet service

```
The FixedVaadinServlet will now ne activated in the moment when an ServletContext with the name `vaadin` exists.
ServletContexts couls be created using a ServletContextHelper.

```
@Component(service = ServletContextHelper.class, scope = ServiceScope.BUNDLE, property = {
		HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_NAME + "=" + VaadinServletContext.NAME,
		HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_PATH + "=/" })
public class VaadinServletContext extends ServletContextHelper {

	@Reference //maybe not longer needed  when places in flow-osgi bundle
	ResourceProvider pr;
	public static final String NAME = "vaadin";
}
```
IF we put this ServletContextHelper into the flow-osgi bundle the VaadinServletContext would be activated by the whiteboard AFTER the immediate activation of theOSGiResourceProvider.

at the moment i just testet the VaadinServletContext inside the starter bundle. but i am quiet sure this could work!

I created the annotation @VaadinContext. this creates ComponentPropertys that prevents from starting with default ServletContext.

Reply:
Interesting solution with `VaadinContext`. 
It might be that we will have to use it later on, but :
- I'm pretty sure that I will find a solution with `ResourceProvider` without this.
- that an additional annotation which one has to know: it means that simplest way of a plain servlet registration won't work. It seems this is a big limitation just to solve one `ResourceProvider` issue. But let's see.

I would like to return back to the my `osgi-bnd`  branch:
thanks a lot for your changes. Now I'm able to understand what's going on there.

What I've got finally:
`-runrequires.flow` - this section will be merged into `-runrequires`. It's not referenced from anywhere. It's just a convenience way to define run requirements instead of having them all together in one `-runrequires`. 
It's just the same as having one `-runrequires` which has all the lines from all the `-runrequires.xxx`. 

That what I was not able to understand and now it's clear.
`-runrequires.flow`  declares additional bundles which `resolve`  was not able to discover as necessary dependencies and we have to declare them explicitly to deploy because without them a Vaadin WAB may not work.

That was the first reason why I've got 404 .

Another reason is `ResourceProvider`  issue.
Now it becomes important.
Indeed if I add 
```
@Reference(cardinality = ReferenceCardinality.MANDATORY)
ResourceProvider pr;
```
then I see a button in the browser.

Without this 404 is shown.

**BUT** the extremely important question is : why it's 404? Why it's not 503 ?
Apparently the only reason is : some exception which is thrown because `ResourceProvider` service is not yet available.
I don't see any exception in the console and 404 instead of 503 is returned.
This is quite severe issue because there is no any way to understand what's wrong : no any error description, no any exception , it just doesn't work.
I strongly dislike this magic.
Do you know why it happens ? May be there is some log which I may check and where the exception is available ?

Reply:
>> That what I was not able to understand and now it's clear.
>> -runrequires.flow declares additional bundles which resolve was not able to discover as necessary dependencies and we >> have to declare them explicitly to deploy because without them a Vaadin WAB may not work.

There would also be a way to have an Annnotation @RequireVaddin (maybe implicet in @VaadinContext) which adds a requirement to the bundle manifest.

And the Capabilities would be provided from vaadin server bundles within the manifest.

That is the same Mechanism why you get a the HTTP-Whiteboard Implementattion while resolving only be adding @HttpWhiteboardServletPattern("/game") to a Servlet.